### PR TITLE
Use limited parallelization and reorder the Makefile

### PR DESCRIPTION
**Pull request by smarterclayton**
Drops support for older constructs, uses explicit parallelization, and avoids fake dependencies by leveraging the `-o` flag for subcommands. Also changes the various test commands to require less conditional logic.

make check is the "simple sanity test"
make test is the full "test" (minus extended, for now)
make test-not-check is the things test runs excluding the things check runs.

Changed build-go to prefer release binaries if they are available.

This is part of fixing #6799 along with openshift/vagrant-openshift#389 - we will run:

    vagrant sync-origin --dont-install
    vagrant test-origin --targets check
    vagrant sync-origin
    vagrant test-origin --targets "test -o check"

For the default jenkins jobs.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@deads2k @stevekuznetsov see the changes here.  This creates a more explicit tree structure using sub make and handles avoiding rerunning phony targets.  Also note the changes to build-go, which ensures that build-go will *not* run once release has been run (so that people can't accidentally revert to testing not-release).  Jordan and I have to sort out the Mac problems, and the upstream drop is just to get the tests to verify they run correctly.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Also, after this, test-integration-docker.sh is removed and there is no reason to manually run test-end-to-end-docker.sh (it will be auto run if you have Docker installed and don't override an ENV var).

File: hack/test-end-to-end.sh [hunk 1]
```diff
@@ -9,6 +9,16 @@ set -o pipefail
 
 STARTTIME=$(date +%s)
 OS_ROOT=$(dirname "${BASH_SOURCE}")/..
+
+if [[ "${TEST_END_TO_END:-}" != "direct" ]]; then
+	if docker version >/dev/null 2>&1; then
+		echo "++ Docker is installed, running hack/test-end-to-end-docker.sh instead."
+		"${OS_ROOT}/hack/test-end-to-end-docker.sh"
```
Comment (liggitt) [inline hack/test-end-to-end.sh]: Need to return or exit here, right?

Comment (smarterclayton) [inline hack/test-end-to-end.sh]: yes

Comment (smarterclayton) [inline hack/test-end-to-end.sh]: I'm just double running to get full coverage ....

**Comment by smarterclayton** [comment]
[test]

File: Makefile [hunk 2]
```diff
@@ -66,91 +55,76 @@ endif
...
+test-unit:
+	TEST_KUBE=1 hack/test-go.sh $(WHAT) $(TESTS) $(TESTFLAGS)
+.PHONY: test-unit
 
-# Build and run the complete test-suite.
-#
-# Args:
-#   GOFLAGS: Extra flags to pass to 'go' when building.
-#   TESTFLAGS: Extra flags that should only be passed to hack/test-go.sh
+# Run integration tests. Compiles its own tests, cannot be run
```
Comment (stevekuznetsov) [inline Makefile]: Can we make any optimization to not need to recompile before integration tests every time?

File: hack/common.sh [hunk 3]
```diff
@@ -124,6 +124,25 @@ os::build::host_platform() {
   echo "$(go env GOHOSTOS)/$(go env GOHOSTARCH)"
 }
 
+# Create a user friendly version of host_platform for end users
+os::build::host_platform_friendly() {
+  platform=${1:-}
```
Comment (stevekuznetsov) [inline hack/common.sh]: `local` everything

File: hack/extract-release.sh [hunk 4]
```diff
@@ -15,10 +15,11 @@ cd "${OS_ROOT}"
 
 # Copy the linux release archives release back to the local _output/local/bin/linux/amd64 directory.
 # TODO: support different OS's?
-os::build::detect_local_release_tars "linux-64bit"
+os::build::detect_local_release_tars $(os::build::host_platform_friendly)
```
Comment (stevekuznetsov) [inline hack/extract-release.sh]: what happens here when `os::build::host_platform_friendly` returns something other than `"linux-64bit"`?

File: Makefile [hunk 5]
```diff
@@ -66,91 +55,76 @@ endif
...
+
+# Run end to end tests. Uses whatever binaries are currently built.
+#
+# Example:
+#   make test-cmd
+test-end-to-end: build
+	hack/test-end-to-end.sh
+.PHONY: test-end-to-end
+
+# Run tools tests.
```
Comment (stevekuznetsov) [inline Makefile]: I'm about to add `hack/test-lib.sh` to test our bash libraries. Does that deserve it's own make target? How do we make that choice?

File: Makefile [hunk 6]
```diff
@@ -181,3 +155,11 @@ release-binaries: clean
 	hack/build-release.sh
 	hack/extract-release.sh
 .PHONY: release-binaries
+
+# Install travis dependencies
+#
+install-travis:
+	hack/install-etcd.sh
```
Comment (deads2k) [inline Makefile]: This should no longer be necessary.

**Comment by deads2k** [comment]
This requires vagrant updates to call the correct targets, right?

File: Makefile (hunk 6 above)
Comment (smarterclayton) [inline Makefile]: Will remove

File: Makefile (hunk 5 above)
Comment (smarterclayton) [inline Makefile]: Yeah, make targets that match naming for everything concrete.

**Comment by smarterclayton** [comment]
This should not change existing behavior of the cluster when merged, the vagrant change would allow us to invoke make targets directly (linked) and then we would change the jenkins jobs to do a two step.  So this can merge without vagrant changes required.

File: hack/extract-release.sh (hunk 4 above)
Comment (smarterclayton) [inline hack/extract-release.sh]: It fails today on any other platform because SERVER doesn't exist.  So same behavior.

File: Makefile (hunk 2 above)
Comment (smarterclayton) [inline Makefile]: I don't think so - I've never tried to a build+install and a go test package build at the same time in one command (I don't think it works).  That would be required to benefit from precompiling integration.

**Comment by smarterclayton** [comment]
Any other comments?

**Comment by smarterclayton** [comment]
[merge]

**Comment by deads2k** [comment]
> Any other comments?

Well timed :)

**Pull request merged**